### PR TITLE
add Dart booster.

### PR DESCRIPTION
**Pull request by marugari**
If we use the Dart booster,
UpdateOneIter calls
  PredictRaw (with dropped trees by DropTrees),
  GetGradient,
  DoBoost (with scaling dropped trees by NormalizeTrees).

### CODE REVIEW

File: src/gbm/gbtree.cc [hunk 1]
```diff
@@ -475,14 +509,277 @@ class GBTree : public GradientBooster {
...
+    GBTree::Save(fo);
+    if (weight_drop.size() != 0) {
+      fo->Write(dmlc::BeginPtr(weight_drop), sizeof(float) * tree_info.size());
+    }
+  }
+
+  void DoBoost(DMatrix* p_fmat,
+               int64_t buffer_offset,
+               std::vector<bst_gpair>* in_gpair) override {
+    const std::vector<bst_gpair>& gpair = *in_gpair;
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: Is this function different from parent's DoBoost? If only difference is CcommitModel, we can make CommitModel virtual and override that function

File: src/gbm/gbtree.cc [hunk 2]
```diff
@@ -475,14 +509,277 @@ class GBTree : public GradientBooster {
...
+  void Configure(const std::vector<std::pair<std::string, std::string> >& cfg) override {
+    GBTree::Configure(cfg);
+    if (trees.size() == 0) {
+      dparam.InitAllowUnknown(cfg);
+    }
+  }
+
+  void Load(dmlc::Stream* fi) override {
+    GBTree::Load(fi);
+    weight_drop.resize(mparam.num_trees);
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: fi->Read(&weight_drop); 

should work

File: src/gbm/gbtree.cc [hunk 3]
```diff
@@ -475,14 +509,277 @@ class GBTree : public GradientBooster {
...
+      CHECK_EQ(fi->Read(dmlc::BeginPtr(weight_drop), sizeof(float) * mparam.num_trees),
+               sizeof(float) * mparam.num_trees)
+          << "Dart: invalid model file";;
+    }
+  }
+
+  void Save(dmlc::Stream* fo) const override {
+    GBTree::Save(fo);
+    if (weight_drop.size() != 0) {
+      fo->Write(dmlc::BeginPtr(weight_drop), sizeof(float) * tree_info.size());
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: fo->Write(weight_drop);

File: src/gbm/gbtree.cc [hunk 4]
```diff
@@ -475,14 +509,277 @@ class GBTree : public GradientBooster {
   std::vector<std::unique_ptr<TreeUpdater> > updaters;
 };
 
+// dart
+class Dart : public GBTree {
+ public:
+  Dart() {
+    num_pbuffer = 0;
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: the initializations was not necessary. vectors are initially empty. num_pbuffer is initialized by constrcuctor of GBTree, this function can be removed

File: doc/parameter.md [hunk 5]
```diff
@@ -74,6 +75,28 @@ Parameters for Tree Booster
 * scale_pos_weight, [default=0]
   - Control the balance of positive and negative weights, useful for unbalanced classes. A typical value to consider: sum(negative  cases) / sum(positive cases) See [Parameters Tuning](param_tuning.md) for more discussion. Also see Higgs Kaggle competition demo for examples: [R](../demo/kaggle-higgs/higgs-train.R ), [py1](../demo/kaggle-higgs/higgs-numpy.py ), [py2](../demo/kaggle-higgs/higgs-cv.py ), [py3](../demo/guide-python/cross_validation.py) 
 
+Additional parameters for Dart
+------------------------------
+* samp_drop [default=0]
+  - type of sampling algorithm.
```
Comment (tqchen) [inline doc/parameter.md]: let use use full name sample_type

File: doc/parameter.md [hunk 6]
```diff
@@ -74,6 +75,28 @@ Parameters for Tree Booster
...
+    - 0: dropped trees are selected uniformly.
+    - 1: dropped trees are selected in proportion to weight.
+* norm_drop [default=0]
```
Comment (tqchen) [inline doc/parameter.md]: normalize_type

Comment (tqchen) [inline doc/parameter.md]: use string(intuitive name) instead of integer for type for the enumeration.

see https://github.com/dmlc/xgboost/blob/master/src/tree/param.h#L74

File: src/gbm/gbtree.cc [hunk 7]
```diff
@@ -47,6 +49,38 @@ struct GBTreeTrainParam : public dmlc::Parameter<GBTreeTrainParam> {
...
+  float rate_drop;
+  /*! \brief whether to drop trees */
+  float skip_drop;
+  /*! \brief learning step size for a time */
+  float learning_rate;
+  // declare parameters
+  DMLC_DECLARE_PARAMETER(DartTrainParam) {
+    DMLC_DECLARE_FIELD(silent).set_default(0)
+        .describe("Not print information during trainig.");
+    DMLC_DECLARE_FIELD(samp_drop).set_default(0)
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: use string instead of type for enumeration. see https://github.com/dmlc/xgboost/blob/master/src/tree/param.h#L74

Internally, the type is still integer, but the enumeration declaration will dispatch the strings into integers, so user can specify string directly

File: src/gbm/gbtree.cc [hunk 8]
```diff
@@ -475,14 +509,277 @@ class GBTree : public GradientBooster {
...
+      this->CommitModel(std::move(new_trees[gid]), gid);
+    }
+  }
+
+  void Predict(DMatrix* p_fmat,
+               int64_t buffer_offset,
+               std::vector<float>* out_preds,
+               unsigned ntree_limit) override {
+    DropTrees(ntree_limit);
+    const MetaInfo& info = p_fmat->info();
```
Comment (tqchen) [inline src/gbm/gbtree.cc]: add a comment that this function is same as GBTree.Predict, but the internal predict function differes.

**Comment by tqchen** [comment]
, I have made a few comments, please let me know when you have new updates, so I can do another round of review.

Thanks for the good job!

**Comment by tqchen** [comment]
@marugari  Any updates on this?

**Comment by marugari** [comment]
Sorry, I'll update by Sunday.

**Comment by tqchen** [comment]
no problem, please take your time
